File: src/scene.js
var Scene = {

    load : function(resourceDir) {
        
        // Setup meshes
        this.eighth.loadMesh(resourceDir + "eighth.obj");
        
        this.half.loadMesh(resourceDir + "half.obj");
        
        this.sphere.loadMesh(resourceDir + "sphere.obj");

        // Setup colors
        this.setupColors();

        // TODO
        var crystal;
        
        crystal = new Crystal(CrystalType.SIMPLE, this.eighth, this.half, this.sphere, this.colors);
        crystal.init();
        this.crystals.push(crystal);
        
        /*crystal = new Crystal(CrystalType.BODY, this.eighth, this.half, this.sphere, this.colors);
        crystal.init();
        this.crystals.push(crystal);
        
        crystal = new Crystal(CrystalType.FACE, this.eighth, this.half, this.sphere, this.colors);
        crystal.init();
        this.crystals.push(crystal);*/
    },

    setupColors : function() {
        this.colors["grey"] = vec3.fromValues(0.5, 0.5, 0.5);
        this.colors["red"] = vec3.fromValues(1.0, 0, 0);
        this.colors["green"] = vec3.fromValues(0, 1.0, 0);
        this.colors["blue"] = vec3.fromValues(0, 0.7, 1.0);
        this.colors["orange"] = vec3.fromValues(1.0, 0.6, 0.2);
        this.colors["black"] = vec3.fromValues(0, 0, 0);
    },

    nextCrystal : function() { 
        this.whichCrystal = (this.whichCrystal + 1) % this.crystals.length;
        this.crystals[this.whichCrystal].setDrawLayers();
    },

    getCrystal : function() { 
        return this.crystals[this.whichCrystal]; 
    },
    
    draw : function(MV, prog) { 
        this.crystals[this.whichCrystal].draw(MV, prog);
    },

    expand : function() {
        for (crystal in this.crystals) {
            crystal.expand();
        }
    },

    contract : function() {
        for (crystal in this.crystals) {
           crystal.contract();
        }
    },

    toggleTranslucency : function() {
        for (crystal in this.crystals) {
            crystal.toggleTranslucency();
        }
    },

    toggleLayers : function() {
        for (crystal in this.crystals) {
            crystal.toggleLayers();
        }
    },

    toggleInspection : function() {
        for (crystal in this.crystals) {
            crystal.toggleInspection();
        }
    },

    isLoaded : function() {
        return this.eighth.isLoaded() && this.half.isLoaded() && this.sphere.isLoaded();
    },
    
    whichCrystal : 0,
    eighth : new Shape(),
    half : new Shape(),
    sphere : new Shape(),
    crystals : [],
    colors : {}
};

File: src/cpp/Scene.cpp
#include <iostream>

#include "Scene.h"
#include "Shape.h"
#include "Program.h"
#include "Crystal.h"
#include "MatrixStack.h"
#include <Eigen/Dense>

using namespace std;
using namespace Eigen;

Scene::Scene() :
    whichCrystal(0)
{
}

Scene::~Scene()
{
}

void Scene::load(const string &RESOURCE_DIR)
{
    eighth = make_shared<Shape>();
    eighth->loadMesh(RESOURCE_DIR + "eighth.obj");
    eighth->init();
    
    half = make_shared<Shape>();
    half->loadMesh(RESOURCE_DIR + "half.obj");
    half->init();

    sphere = make_shared<Shape>();
    sphere->loadMesh(RESOURCE_DIR + "sphere.obj");
    sphere->init();

    crystals.push_back(make_shared<Crystal>(Crystal::SIMPLE, eighth, half, sphere));
    crystals.back()->init();
    
    crystals.push_back(make_shared<Crystal>(Crystal::BODY, eighth, half, sphere));
    crystals.back()->init();
    
    crystals.push_back(make_shared<Crystal>(Crystal::FACE, eighth, half, sphere));
    crystals.back()->init();
}

void Scene::nextCrystal()
{
    whichCrystal = (whichCrystal + 1) % crystals.size();

    // Make sure it starts as falling layers
    crystals[whichCrystal]->setDrawLayers();
}


shared_ptr<Crystal> Scene::getCrystal()
{
    return crystals[whichCrystal];
}

void Scene::draw(shared_ptr<MatrixStack> MV, const shared_ptr<Program> prog)
{
    crystals[whichCrystal]->draw(MV, prog);
}

void Scene::expand()
{
    for (unsigned int i = 0; i < crystals.size(); i++) {
        crystals[i]->expand();
    }
}

void Scene::contract()
{
     for (unsigned int i = 0; i < crystals.size(); i++) {
         crystals[i]->contract();
     }
}

void Scene::toggleTranslucency()
{
    for (unsigned int i = 0; i < crystals.size(); i++) {
         crystals[i]->toggleTranslucency();
    }
}

void Scene::toggleLayers()
{
    for (unsigned int i = 0; i < crystals.size(); i++) {
         crystals[i]->toggleLayers();
    }
}

void Scene::toggleInspection()
{
    for (unsigned int i = 0; i < crystals.size(); i++) {
         crystals[i]->toggleInspection();
    }
}
